Skeleton for cluster lens simulations

furcelay opened the pull request:
This is the first pull following #134 and provides a skeleton of the classes described in that issue. The classes are empty except for `ClusterLens(LensBase)` which is partially implemented. 

`ClusterLens` takes the parameters that describe a halo (by the time only EPL) and a table with the parameters of each subhalo and generates the Lenstronomy keywords to simulate the lens. This is exemplified in a new notebook `cluster_lens.ipynb`.

nkhadka21 reviewed (changes requested):
Hi @furcelay , thank you very much for your PR. I have some major comments. GalaxyClusterLenses and ClusterLensPop are mostly duplicates of Lens and LensPop classes. I think you should start from the small steps. I encourage you to start from generating cluster deflector and sources for cluster lenses. and then, you can think about ClusterLens and ClusterLensPop (might be able to use existing classes with addition of some extra parameters). Please let me know if you have any questions!

furcelay commented on slsim/Pipelines/cluster_pipeline.py:
```diff
@@ -0,0 +1,3 @@
+class ClusterPipeline:
```
Removed this class as it is not needed yet.

furcelay commented on slsim/cluster_lens_pop.py:
```diff
@@ -0,0 +1,105 @@
+from slsim.lensed_population_base import LensedPopulationBase
+
+
+class ClusterLensPop(LensedPopulationBase):
```
Removed this class

furcelay commented on tests/test_Pipelines/test_cluster_pipeline.py:
```diff
@@ -0,0 +1,14 @@
+from slsim.Pipelines.cluster_pipeline import ClusterPipeline
+
+
+class TestClusterPipeline(object):
```
Removed this class

furcelay commented on tests/test_cluster_lens.py:
```diff
@@ -0,0 +1,233 @@
+
+
+class TestClusterLens(object):
+    # pytest.fixture(scope='class')
```
Removed this class

furcelay commented on slsim/cluster_lens.py:
```diff
@@ -0,0 +1,714 @@
+import warnings
+
+
+class ClusterLens(LensedSystemBase):
```
I've removed the ClusterLens class and implemented `NFWCluster` which is a new `Deflector` subclass to manage clusters. It can work with the existing `Lens` class by modifying the `deflector_mass_model_lenstronomy`. Another way might be to handle the mass model in the `Deflector` class and just call it from the `Lens` class (see #156).

furcelay commented:
I've removed the redundant or unneeded classes and implemented a new Deflector to handle clusters. This deflector is a NFW halo without a luminous component and subhalos that can be any other deflector (EPL+Sérsic or NFW+Hernquist). I've also implemented a new function to compute its line-of-sight velocity dispersion of the halo following [Lokas & Mamon 2001](https://arxiv.org/abs/astro-ph/0002395) as it cannot be weighted by the light as in the NFW + Hernquist model.

furcelay commented on slsim/Deflectors/DeflectorTypes/nfw_cluster.py:
```diff
@@ -0,0 +1,66 @@
+
+        m_halo, c_halo = self.halo_properties
+        rs_arcsec, _ = lens_cosmo.nfw_physical2angle(m_halo, c_halo)
+        vel_disp = vel_disp_nfw_aperture(
```
Hi @sibirrer, indeed the velocity dispersion is the one of the dark matter particles without any light tracer. This is because the velocity dispersion of the dark matter halo of groups or clusters cannot be measured in the same way as galaxies. It is usually measured by the spread in the redshift of the subhalos within a radius and assuming the radial profile for the velocity dispersion from Łokas & Mamon 2001. So it should be consistent with the measured velocity dispersion with some considerations on the aperture radius.

sibirrer reviewed (approved):
Thank you @furcelay! I leave it to @nkhadka21 for the final review and merge.

nkhadka21 commented on slsim/Deflectors/DeflectorTypes/nfw_cluster.py:
```diff
@@ -0,0 +1,66 @@
+    - 'z': redshift of deflector
+    """
+
+    def __init__(self, deflector_dict, subhalos_list):
```
Please add doc string for deflector_dict and subhalos_list. Clarify what subhalos_list should contain. Might be useful to explain how halo_EPl.fits and subhaloes_table.fits can be used in this class.

furcelay replied:
Hi @nkhadka21, I've added the docstring for that method. Also included the fits files in `test_nfw_cluster` to exemplify their use, both have an astropy table with the quantities to init the halo and subhalos respectively.

nkhadka21 reviewed (changes requested):
Hi @furcelay , thank you very miuch for this update! Looks very good. I have a minor comment. Please address it.

nkhadka21 reviewed (approved):
@furcelay, thank you very much for the update! I approved it and excited to see your progress on other aspects of the cluster lens code!

furcelay commented:
Thank you! I hope to have more progress soon.
